### DISCUSSION TITLE
Dropouts on esp32-player-sd-a2dp for Arduino player

### DISCUSSION THREAD

**Discussion by frankcohen** [pschatzmann/arduino-audio-tools#852] [Q&A] [answered]
Hi, thank you x 100 for your great libraries. I created a fun air guitar project. It's an ESP32 SD A2DP player. It has a NAND SD chip and accelerometer to sense what the user is doing... move the board and it plays the music on a Bluetooth speaker.

The code is GPL v3 and at https://github.com/frankcohen/ReflectionsOS/tree/main/Experiments/AirGuitar

A video showing it playing is at https://youtu.be/J06CSqpnano

I pegged the audio to play continuously while testing it. I found a lot of drop-outs when playing the music. I tested the SD memory (it runs over SPI) and I don't see lags or latency. A snippet of my code is here:

BluetoothA2DPSource a2dp_source;
. . .
int32_t get_sound_data(Channels* data, int32_t len) {

    int32_t result_len = 0;
    if ( playing )
    {
      // the data in the file must be in int16 with 2 channels 
      size_t result_len_bytes = sound_file.read((uint8_t*)data, len * frame_size_bytes );
      // result is in number of frames
      result_len = result_len_bytes / frame_size_bytes;
      // ESP_LOGD("get_sound_data", "%d -> %d",len );
      totalmusic += result_len_bytes;
    }

    return result_len;    
}
. . .
  a2dp_source.start("JBL Charge 4", get_sound_data);  

Do you notice something I'm doing wrong?

-Frank

**Comment by pschatzmann** [comment, marked as answer]
You don't really provide enough information. In general I recommend not to use the stream based implementation, but the one using the callback directly that you can find in the basic examples.

**Comment by frankcohen** [comment]
Thank you! I’ll look at the basic examples for a non-streaming direct callback. I was using these examples:

https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-basic-api/base-player-a2dp/base-player-a2dp.ino
https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-player/player-sdfat-a2dp/player-sdfat-a2dp.ino
https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-player/player-sd-i2s/player-sd-i2s.ino

I want to avoid SDFat. On the ESP32 and with the NAND SD chip I find SDfat to be overkill.

Is there a specific example I should use? I didn’t find a esp32-player-sd-a2dp one. Glad to contribute one.

-Frank

**Comment by pschatzmann** [comment]
If you want to use the player I suggest that you read the corresponding Wiki: https://github.com/pschatzmann/arduino-audio-tools/wiki/The-Audio-Player-Class
